src/main/scala/viper/gobra/frontend/Desugar.scala

```diff
@@ -5089,34 +5165,62 @@ object Desugar extends LazyLogging {
   /**
     * Capabilities to store specification relevant to package initialization across
     * multiple desugarers. In particular, it provides functionality to store and retrieve
-    * preconditions for all imports of a package and the postconditions of all programs of
-    * a package.
+    * preconditions for all imports of a package, the package invariants of all packages,
+    * and the friend clauses.
     */
   private class PackageInitSpecCollector {
+    private var nonDupPkgInvs: Map[PPackage, Vector[in.Assertion]] = Map.empty
+    private var dupPkgInvs: Map[PPackage, Vector[in.Assertion]] = Map.empty
+
+    // Register the invariants of pkg. Each invariant is accompanied by a bool that
+    // is true iff the invariants are meant to be duplicable
+    def registerPkgInvariants(pkg: PPackage, dInvs: Vector[(in.Assertion, Boolean)]) = {
+      assert(!nonDupPkgInvs.contains(pkg) && !dupPkgInvs.contains(pkg))
+      val (dups, nonDups) = dInvs.partitionMap { inv =>
+        if (inv._2) Left(inv._1) else Right(inv._1)
+      }
+      dupPkgInvs += pkg -> dups
+      nonDupPkgInvs += pkg -> nonDups
+    }
+
+    def getNonDupPkgInvariants(): Map[PPackage, Vector[in.Assertion]] = nonDupPkgInvs
+    def getDupPkgInvariants(): Map[PPackage, Vector[in.Assertion]] = dupPkgInvs
```

Comment:
Wouldn't it be sufficient to simply store `PPackage`s and compute (& if necessary memoize) `nonDupPkgInvs` and `dupPkgInvs`? It seems this would result in a much simpler interface and also make clients easier to read by encapsulating more functionality in `PackageInitSpecCollector`

Reply:
I think this question is very related to the first one (if we move `registerPkgInitData` to the `PackageInitSpecCollector` class, we end up effectively doing this). Once again, the reason I didn't do this is that I wanted `PackageInitSpecCollector` to act only as a store of simple metadata, instead of having more complicated desugaring logic here.